crates/ide/src/navigation_target.rs

```diff
@@ -167,10 +167,14 @@ impl NavigationTarget {
     }
 }
 
-impl ToNav for FileSymbol {
-    fn to_nav(&self, db: &RootDatabase) -> NavigationTarget {
-        NavigationTarget {
-            file_id: self.file_id,
+impl TryToNav for FileSymbol {
+    fn try_to_nav(&self, db: &RootDatabase) -> Option<NavigationTarget> {
+        let semantics = Semantics::new(db);
+        let full_range = self.loc.original_range(&semantics)?;
+        let name_range = self.loc.original_name_range(&semantics)?;
+
```

Comment:
This feels wrong, semantics is only really needed because of the caching behaviour, but we don't really need caching here at all since we discard the semantics after this again. Since this is the only call site for these functions it should be fine to accept RootDatabase in the functions instead.

Reply:
I agree with you, however there is some complication with this. So, the whole "make a semantics and throw it away" is already canon in the codebase: 

https://github.com/rust-analyzer/rust-analyzer/blob/92abc56bc928fb2a11b8f5b2a37e3c9ee31102d7/crates/ide/src/parent_module.rs#L28
https://github.com/rust-analyzer/rust-analyzer/blob/92abc56bc928fb2a11b8f5b2a37e3c9ee31102d7/crates/ide/src/parent_module.rs#L59
https://github.com/rust-analyzer/rust-analyzer/blob/3b3063fb2d1804b3dc42184027ef08073de03fd9/crates/ide/src/goto_declaration.rs#L17

etc...

But the real reason is because of https://github.com/rust-analyzer/rust-analyzer/pull/10872/files#diff-596f08c5fa45153f5a0be64c786ae5e1351e22e68d45563a653b7a33965e0e77R137

Notably, if we don't expand_or_parse the file id in semantics, then NameClass::classify will panic (because the thing is not in the cache.)

So, given this, I decided to go over this approach. If we made it take RootDatabase instead of semantics, then would have to special case the thing in ItemLocator and expose the internal  hirfileid/ptr's in DeclarationLocation. So this seemed like the least messy, and it seems like "semantics for throwaway" seems to be an OK pattern already.

Reply:
> it seems like "semantics for throwaway" seems to be an OK pattern already.

Commenting just on this, I'm not sure it is. We sometimes change the code to reuse those. Throwing them away is still a common pattern but it probably shouldn't be.

Reply:
Fair. A pair of fresh eyes has made me realize that the requirement in semantics is not needed for anything but extracting the syntax node, so I did just that here: b0c7ff39b8277aaf80c93a83200dc732792bac7d